src/util/pointabilities.cpp

```diff
@@ -0,0 +1,143 @@
+/*
+Minetest
+Copyright (C) 2023 cx384
+
+This program is free software; you can redistribute it and/or modify
+it under the terms of the GNU Lesser General Public License as published by
+the Free Software Foundation; either version 2.1 of the License, or
+(at your option) any later version.
+
+This program is distributed in the hope that it will be useful,
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+GNU Lesser General Public License for more details.
+
+You should have received a copy of the GNU Lesser General Public License along
+with this program; if not, write to the Free Software Foundation, Inc.,
+51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+*/
+
+#include "pointabilities.h"
+
+#include "serialize.h"
+#include "exceptions.h"
+#include <sstream>
+
+PointabilityType Pointabilities::deSerializePointabilityType(std::istream &is)
+{
+	PointabilityType pointable_type = static_cast<PointabilityType>(readU8(is));
+	switch(pointable_type) {
+		case PointabilityType::POINTABLE:
+		case PointabilityType::POINTABLE_NOT:
+		case PointabilityType::POINTABLE_BLOCKING:
+			break;
+		default:
+			pointable_type = PointabilityType::POINTABLE;
+			break;
+    }
+    return pointable_type;
+}
+
+void Pointabilities::serializePointabilityType(std::ostream &os, PointabilityType pointable_type)
+{
+	writeU8(os, static_cast<u8>(pointable_type));
+}
+
+std::string Pointabilities::toStringPointabilityType(PointabilityType pointable_type)
+{
+	switch(pointable_type) {
+		case PointabilityType::POINTABLE:
+			return "true";
+		case PointabilityType::POINTABLE_NOT:
+			return "false";
+		case PointabilityType::POINTABLE_BLOCKING:
+			return "\"blocking\"";
+		default:
+			return "unknown";
```

Comment:
How can this happen?

Reply:
A server could add a new `PointabilityType` that an old client does not know about.
Then the client somehow called `dump(obj)` with CSM and gets `... pointable=unknown ...`
(I haven't tested this and I'm not completely sure if this can actually happen currently, but in the future it may be possible.)

Reply:
I think ideally a new server shouldn't be sending values to old clients that don't understand them. The server should already decide what to send to old clients; the client doesn't need to handle it.

Reply:
It's defaulted to `POINTABLE` in case the type is unknown.
(This is future proof, since compatibility is already checked with a version filed in the item deserialization functions. So it will only happen if a client does not behave well. As in all other places that default invalid values, no error/warning message will happen.)
I also moved `return "unknown";` out of the switch, normal it would never be executed, but completely removing it would result in a compiler warning.

Reply:
> but completely removing it would result in a compiler warning.

I usually deal with this by adding a `throw std::logic_error("invalid enum value");`, but I can live with this.